pkg/controllers/applicationfailover/rb_application_failover_controller.go

```diff
@@ -0,0 +1,249 @@
+				klog.Error(err)
+				return err
+			}
+		} else {
+			binding.Spec.RemoveCluster(cluster)
+		}
+	}
+
+	return nil
+}
+
+func (c *RBApplicationFailoverController) updateBinding(binding *workv1alpha2.ResourceBinding, allClusters map[string]struct{}, needEvictClusters []string) error {
+	if err := c.Update(context.TODO(), binding); err != nil {
+		for _, cluster := range needEvictClusters {
+			helper.EmitClusterEvictionEventForResourceBinding(binding, cluster, c.EventRecorder, err)
+		}
+		klog.ErrorS(err, "Failed to update binding", "binding", klog.KObj(binding))
+		return err
+	}
+	for _, cluster := range needEvictClusters {
+		delete(allClusters, cluster)
+	}
+	if !features.FeatureGate.Enabled(features.GracefulEviction) {
+		for _, cluster := range needEvictClusters {
+			helper.EmitClusterEvictionEventForResourceBinding(binding, cluster, c.EventRecorder, nil)
+		}
+	}
+
+	return nil
+}
+
+// SetupWithManager creates a controller and register to controller manager.
+func (c *RBApplicationFailoverController) SetupWithManager(mgr controllerruntime.Manager) error {
+	c.workloadUnhealthyMap = newWorkloadUnhealthyMap()
+	resourceBindingPredicateFn := predicate.Funcs{
+		CreateFunc: func(createEvent event.CreateEvent) bool {
+			obj := createEvent.Object.(*workv1alpha2.ResourceBinding)
+			if obj.Spec.Failover == nil || obj.Spec.Failover.Application == nil {
+				return false
+			}
+
+			if len(obj.Status.AggregatedStatus) == 0 {
+				return false
+			}
+
+			resourceKey, err := helper.ConstructClusterWideKey(obj.Spec.Resource)
+			if err != nil {
+				// Never reach
+				klog.Errorf("Failed to construct clusterWideKey from binding(%s/%s)", obj.Namespace, obj.Name)
+				return false
+			}
+
+			if !c.ResourceInterpreter.HookEnabled(resourceKey.GroupVersionKind(), configv1alpha1.InterpreterOperationInterpretHealth) {
+				return false
+			}
+
+			if !obj.Spec.PropagateDeps {
+				return false
+			}
+
+			return true
+		},
+		UpdateFunc: func(updateEvent event.UpdateEvent) bool {
+			oldObj := updateEvent.ObjectOld.(*workv1alpha2.ResourceBinding)
+			newObj := updateEvent.ObjectNew.(*workv1alpha2.ResourceBinding)
+
+			if (oldObj.Spec.Failover == nil || oldObj.Spec.Failover.Application == nil) &&
+				(newObj.Spec.Failover == nil || newObj.Spec.Failover.Application == nil) {
+				return false
+			}
+
+			if len(newObj.Status.AggregatedStatus) == 0 {
+				return false
+			}
+
+			resourceKey, err := helper.ConstructClusterWideKey(newObj.Spec.Resource)
+			if err != nil {
+				// Never reach
+				klog.Errorf("Failed to construct clusterWideKey from binding(%s/%s)", newObj.Namespace, newObj.Name)
+				return false
+			}
+
+			if !c.ResourceInterpreter.HookEnabled(resourceKey.GroupVersionKind(), configv1alpha1.InterpreterOperationInterpretHealth) {
+				return false
+			}
+
+			if !newObj.Spec.PropagateDeps {
+				return false
+			}
+			return true
+		},
+		DeleteFunc:  func(deleteEvent event.DeleteEvent) bool { return false },
```

Comment:
Why use `return false` here?

Reply:
For this controller, we only deal with the `update` event which updates the `deleteTimeStamp` in the `ResourceBinding`.

Reply:
I guess because @Poor12 thought there is nothing to do for the delete event. 

In my opinion, we should clean up the `workloadUnhealthyMap` at 
```
	if err := c.Client.Get(ctx, req.NamespacedName, binding); err != nil {
		if apierrors.IsNotFound(err) {
			return controllerruntime.Result{}, nil
		}
		return controllerruntime.Result{Requeue: true}, err
	}
```
instead of `if !binding.DeletionTimestamp.IsZero()`.

Reply:
> I am a little confused about the benefit between these two solutions.

The state of `DeletionTimestamp` is transitory, the controller might miss the event. Take RB deletion process as an example:
```
time1: start to delete the RB, requeue the RB after setting the `DeletionTimestamp`.
time2: RB is deleted, requeue the RB.
time3: The controller starts to reconcile the RB, it can't get the RB anymore.
```

Reply:
> I guess because @Poor12 thought there is nothing to do for the delete event.
> 
> In my opinion, we should clean up the `workloadUnhealthyMap` at
> 
> ```
> 	if err := c.Client.Get(ctx, req.NamespacedName, binding); err != nil {
> 		if apierrors.IsNotFound(err) {
> 			return controllerruntime.Result{}, nil
> 		}
> 		return controllerruntime.Result{Requeue: true}, err
> 	}
> ```
> 
> instead of `if !binding.DeletionTimestamp.IsZero()`.

I find that we cannot deal with it in the above logic. If apierrors.IsNotFound(err) == true which means that binding has been deleted, actually we cannot Get the resource of binding so that we do not know how to cleanup `workloadUnhealthyMap`.